core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,415 @@
+            requested--;
+            if (!publishedAny) {
+                publishedAny = true;
+            }
+
+            publish();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (closed) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            publishDownstream(item);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            close(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            setState(State.REQUIRE_BODY, State.REQUIRE_TAIL);
+            publish();
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (notifyCancellation) {
+                downstream.onError(cause);
+            }
+            downstream = NoopSubscriber.get();
+            completionFuture.completeExceptionally(cause);
+            release();
+        }
+
+        private void close(@Nullable Throwable cause) {
+            if (executor.inEventLoop()) {
+                close0(cause);
+            } else {
+                executor.execute(() -> close0(cause));
+            }
+        }
+
+        private void close0(@Nullable Throwable cause) {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            if (cause == null) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                downstream.onError(cause);
+                completionFuture.completeExceptionally(cause);
+            }
+            release();
+        }
+
+        private void release() {
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+                upstream.cancel();
+            }
+            if (head != null) {
+                PooledObjects.close(head);
+            }
+            if (tail != null) {
+                PooledObjects.close(tail);
+            }
+        }
+
+        private boolean setState(State oldState, State newState) {
+            assert newState != State.REQUIRE_HEAD : "oldState: " + oldState + ", newState: " + newState;
+            return stateUpdater.compareAndSet(this, oldState, newState);
```

Comment:
We don't need CAS operation if it is executed in a single thread.

Reply:
✅ updated!